packages/sdk-js/src/feature-repository.ts

```diff
@@ -0,0 +1,346 @@
+  clearAutoUpdates();
+
+  if (globalThis.localStorage) {
+    try {
+      globalThis.localStorage.removeItem(cacheSettings.localStorageKey);
+    } catch (e) {
+      // Ignore localStorage errors
+    }
+  }
+}
+
+function getApiHostAndKey(instance: GrowthBook): [ApiHost, ClientKey, boolean] {
+  // eslint-disable-next-line
+  const ctx = (instance as any).context as Context;
+  const apiHost = ctx.apiHost || "";
+  const clientKey = ctx.clientKey || "";
+  return [apiHost.replace(/\/*$/, ""), clientKey, !!ctx.enableDevMode];
+}
+
+export async function primeCache(
+  instance: GrowthBook
+): Promise<FeatureApiResponse | null> {
+  const [apiHost, clientKey, enableDevMode] = getApiHostAndKey(instance);
+  if (!clientKey) return null;
+  const key: RepositoryKey = `${apiHost}||${clientKey}`;
+  initializeCache();
+  const existing = cache.get(key);
+  if (existing && !enableDevMode) {
+    // Reload features in the backgroud if stale
+    if (existing.staleAt < new Date()) {
+      fetchFeatures(apiHost, clientKey);
+    }
+    // Otherwise, if we don't need to refresh now, start a background sync
+    else if (backgroundSyncEnabled) {
+      startAutoUpdate(apiHost, clientKey, key);
+    }
+    return existing.data;
+  } else {
+    const data = await fetchFeatures(apiHost, clientKey);
+    return data;
+  }
+}
+
+export async function loadFeatures(
+  instance: GrowthBook,
+  options: LoadFeaturesOptions = {}
+) {
+  const [apiHost, clientKey] = getApiHostAndKey(instance);
+  if (!clientKey) return [null, null];
+  const key: RepositoryKey = `${apiHost}||${clientKey}`;
+  if (!key) return;
+
+  // Fetch features with an optional timeout
+  const data = await promiseTimeout(primeCache(instance), options.timeout);
+
+  if (data) {
+    await setFeaturesOnInstance(instance, data);
+  }
+  if (options.autoUpdate) {
+    subscribe(key, instance);
+  }
+}
+export function unsubscribe(instance: GrowthBook) {
+  subscribedInstances.forEach((s) => s.delete(instance));
+}
+
+// Private functions
+
+// Guarantee the promise always resolves within {timeout} ms
+// Resolved value will be `null` when there's an error or it takes too long
+// Note: The promise will continue running in the background, even if the timeout is hit
+function promiseTimeout<T>(
+  promise: Promise<T>,
+  timeout?: number
+): Promise<T | null> {
+  return new Promise((resolve) => {
+    let resolved = false;
+    // eslint-disable-next-line
+    let timer: any;
+    const finish = (data?: T) => {
+      if (resolved) return;
+      resolved = true;
+      timer && clearTimeout(timer);
+      resolve(data || null);
+    };
+
+    if (timeout) {
+      timer = setTimeout(() => finish(), timeout);
+    }
+
+    promise.then((data) => finish(data)).catch(() => finish());
+  });
+}
+
+// Subscribe a GrowthBook instance to feature changes
+function subscribe(key: RepositoryKey, instance: GrowthBook) {
+  const subs = subscribedInstances.get(key) || new Set();
+  subs.add(instance);
```

Comment:
maybe check if instance is already in subs first before adding, in case a GB is double-subscribed somehow. not sure how that could happen, but just to be defensive...

Reply:
Sets automatically remove duplicates

Reply:
Opps for some reason I was thinking this was an array